api/agent/protocol/http.go

```diff
@@ -33,6 +37,10 @@ func (h *HTTPProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) e
 
 	resp, err := http.ReadResponse(bufio.NewReader(h.out), ci.Request()) // TODO timeout
 	if err != nil {
+		// From here (boo stdlib): https://github.com/golang/go/blob/d3c1df712658398f29bc8bebd6767e7b3cac2d12/src/net/http/response.go#L174
```

Comment:
I don't think we should cherry pick based on error string match. There are many error variations in the stdlib (3-4 cases of malformed http only) so I think we should do one of the following:

1) mask all errors coming from ReadResponse as "invalid response from function" (hides internals, but better than generic 500)
2) pass through all errors from  ReadResponse inside NewAPIError (more user friendly, but could expose internal details)

Should these errors be 502 also?

IMHO, I'd pick '1' but using 502 bad gateway. (Since we are using 504 for timeout already and acting as a gw.)

Reply:
at least one exception is that if the pipe closes they'll get io.ErrClosedPipe, but in theory that error won't get plumbed up since it's past the timeout (at some point, it may on accident leak out). 

i was hoping for a typed error from stdlib, sadly it's not the case. 

502 makes sense, 1) makes sense. we can log the real error. the error we get back conveniently does have some info in it such as `invalid HTTP status code: "MYOOPSYLOGLINE"` which could tip a user off as to what they're doing wrong. 

I was thinking it's possible since we're basically only doing i/o that we could plumb up any error from Dispatch as a `models.APIError` but the surface area may be too large. it would cover the json case nicely and also cover `req.Write` failures, but it's less clear what the exact error should be. something along the lines of "container i/o failed".

Reply:
"invalid response from function", "cannot process function response" or something like that maybe? "container i/o failed" sounds good too except maybe that sounds like it could also be fn server issue.

Reply:
I had another thought about this the other day that in http / json we could probably handle cases where uses write accidental log lines to stdout before the actual http / json blob. i'm not sure how worth it that is, but it's definitely a lego in the foot at the moment. it may well be better to try to get them to correct their behavior than covering up their misbehaving. 'invalid function response' kind of only makes sense in the case handled in this patch.

Reply:
agree with pt.1 - if function writes bad http response (whatever except well-structured http response) Fn should be more verbose for this cases and i'd like to get "invalid response from function" because it also indicates broken FDK as well as a user's faults.

Reply:
it's more a matter of where to put it at this point. if we catch every error and say that it's not necessarily true. we're running on a hijacked connection, any number of things could go wrong there that aren't the functions fault, hence checking for very specific errors.